### PR TITLE
feat(bulk-load): bulk load ingestion part1 - meta start ingestion

### PR DESCRIPTION
**Pull request by hycdong**
The whole bulk load ingestion process is like:
1. **meta server set app and all partitions bulk load status as ingesting**
2. meta send two requests to primary 
    - **send ingestion request to primary**
    - send bulk load request to query ingestion status #457
3. primary handle two requests above:
    - primary consider ingestion request as a client write and replica group execute 2pc process to do actual rocksdb ingestion operation
    - primary handle bulk load request and broadcast it to secondary
4. secondary handle two cases:
    - ingestion files through 2pc
    - report ingestion status to primary 
5. primary report group ingestion status to primary
6. meta handle two response
    - execute error handling from ingestion_response
    - handle group ingestion status and do bulk load status transformation

This pull request is about meta start ingestion, step 1 and ingestion request of step 2.
- update meta bulk load status
    - when bulk load downloading stage finished, each partition bulk load status and app status are downloaded, then update each partition bulk load status from downloaded to ingesting (in `on_partition_bulk_load_reply`)
    - when all partitions updated, update bulk load status from downloaded to ingesting (in `update_partition_status_on_remote_storage_reply`)
    - when app status updated, send ingestion request to each primary (in `update_app_status_on_remote_storage_reply`)
- send ingestion request (in `partition_ingestion`)
    - this function will execute in default thread pool
    - add related thrift structure and task code
    - replica should consider ingestion request as a write request, as a result, meta server should create a request which is like sent from user client. Client requests have a feature that `msg->header.gpid = pid`. 

**_Special explanation and discussion_**:
I don't use `rpc_holder` utility for meta server to send ingestion request to primary, because I need to update gpid filed in message header, and it is a special case for meta to create a client request. If you have any suggestion about it, please comment this pull request or discuss with me offline.

### CODE REVIEW

File: include/dsn/tool-api/rpc_message.h [hunk 1]
```diff
@@ -175,6 +175,11 @@ class message_ex : public ref_counter,
                                                        int size,
                                                        int thread_hash = 0,
                                                        uint64_t partition_hash = 0);
+    // Used for bulk load ingestion
+    // meta send bulk load ingestion request to primary, primary will consider it as a special write
+    // request, in other word, create a request which will be considered sent from user client
```
Comment (levy5307) [inline include/dsn/tool-api/rpc_message.h:180]: in other words
considered as

File: src/dist/replication/meta_server/meta_bulk_load_service.cpp [hunk 2]
```diff
@@ -394,7 +394,8 @@ void bulk_load_service::on_partition_bulk_load_reply(error_code err,
         handle_app_downloading(response, primary_addr);
         break;
     case bulk_load_status::BLS_DOWNLOADED:
-        handle_app_downloaded(response);
+        update_partition_status_on_remote_storage(
+            response.app_name, response.pid, bulk_load_status::BLS_INGESTING);
         // when app status is downloaded or ingesting, send request frequently
```
Comment (levy5307) [inline src/dist/replication/meta_server/meta_bulk_load_service.cpp:399]: why app status is downloaded or ingesting, send request will be more frequently?

Comment (hycdong) [reply src/dist/replication/meta_server/meta_bulk_load_service.cpp:399]: When app status is downloaded and ingesting, partition status may be ingesting. 
When partition status is updated to ingesting, replica will reject other write requests and recover when primary notice group finish ingesting. Primary will check if group finish ingesting while receive bulk_load_request. As a result, meta send request frequently will let replica recovering write requests more quickly.

File: src/dist/cmd/command_types.cpp [hunk 3]
```diff
@@ -149,6 +149,6 @@ void command::printTo(std::ostream &out) const
         << "arguments=" << to_string(arguments);
     out << ")";
 }
-} // namespace cmd
-} // namespace dist
-} // namespace dsn
+}
+}
```
Comment (acelyc111) [inline src/dist/cmd/command_types.cpp:153]: Why these auto generated code has been removed?

File: include/dsn/tool-api/rpc_message.h [hunk 4]
```diff
@@ -175,6 +175,11 @@ class message_ex : public ref_counter,
                                                        int size,
                                                        int thread_hash = 0,
                                                        uint64_t partition_hash = 0);
+    // Used for bulk load ingestion
+    // meta send bulk load ingestion request to primary, primary will consider it as a special write
+    // request, in other words, create a request which will be considered as sent from user client
+    DSN_API static message_ex *
+    create_client_request(dsn::task_code rpc_code, dsn::gpid pid, int timeout_milliseconds = 0);
```
Comment (neverchanje) [inline include/dsn/tool-api/rpc_message.h:182]: If this is for bulk-load only, put it on replica/bulkload folder. Don't add upper-level details in a low-level module.

File: src/dist/replication/replication.thrift [hunk 5]
```diff
@@ -1003,6 +1003,18 @@ struct group_bulk_load_response
...
 
+struct ingestion_request
+{
+    1:string                app_name;
+    2:bulk_load_metadata    metadata;
+}
+
+struct ingestion_response
+{
+    1:dsn.error_code    err;
```
Comment (neverchanje) [inline src/dist/replication/replication.thrift:1014]: Any "Possible errors" comments?

File: src/dist/replication/replication.thrift [hunk 6]
```diff
@@ -1003,6 +1003,18 @@ struct group_bulk_load_response
     3:partition_bulk_load_state bulk_load_state;
 }
 
+struct ingestion_request
```
Comment (neverchanje) [inline src/dist/replication/replication.thrift:1006]: Any "from client to replica-server"-like comments?

File: include/dsn/tool-api/rpc_message.h (hunk 4 above)
Comment (neverchanje) [reply include/dsn/tool-api/rpc_message.h:182]: BTW. This function seems touching no private member. Why not make it a global&static function?

File: src/dist/cmd/command_types.cpp (hunk 3 above)
Comment (hycdong) [reply src/dist/cmd/command_types.cpp:153]: It seems that code auto generated by thrift won't add namespace comments at the end of the file. You can see other files generated by thrift, such as `include/dsn/dist/replication/replication_types.h`

File: include/dsn/tool-api/rpc_message.h (hunk 4 above)
Comment (hycdong) [reply include/dsn/tool-api/rpc_message.h]: I remove this function.

**Pull request merged**